rockcraft/layers.py

```diff
@@ -45,14 +47,53 @@ def archive_layer(
     candidates = _gather_layer_paths(new_layer_dir, base_layer_dir)
     layer_paths = _merge_layer_paths(candidates)
 
-    with tarfile.open(temp_tar_file, mode="w") as tar_file:
-        # Iterate on sorted keys, so that the directories are always listed before
-        # any files that they contain (otherwise tools like Docker might choke on
-        # the layer tarball).
-        for arcname in sorted(layer_paths):
+    if not temp_tar_file.parent.exists():
+        temp_tar_file.parent.mkdir(parents=True)
+
+    layer_contents: list[Path] = []
+    with TemporaryDirectory() as tmpdir:
+        tmppath = Path(tmpdir)
+        # Just create an empty tar file if there's nothing to put in the layer
+        # GNU tar simply refuses to create an actual empty tar file though, so just use
+        # python's tarfile
+        if not layer_paths:
+            with tarfile.open(temp_tar_file, "w"):
+                pass
+            return
+
+        # Walk in reverse to avoid encountering not-yet created dirs
+        for arcname in sorted(layer_paths, reverse=True):
             filepath = layer_paths[arcname]
-            emit.debug(f"Adding to layer: {filepath} as '{arcname}'")
-            tar_file.add(filepath, arcname=arcname, recursive=False)
+            emit.debug(f"Adding to layer: {filepath} as {arcname!r}")
+
+            # Construct a new file at `arcname` with the same contents as `filepath`.
+            # This emulates the `arcname` parameter of `tarfile.open()`, which is not
+            # present in GNU tar.
+            new_path = tmppath / arcname
+            layer_contents.append(new_path.relative_to(tmppath))
+            if new_path.is_dir() and new_path.exists():
+                continue
+            new_path.parent.mkdir(parents=True, exist_ok=True)
+            filepath.rename(new_path)
```

Comment:
This whole piece if confusing me. Despite the comment I am very unsure what it is doing. By recreating directories are we not risking losing permissions/attributes on them?